### ISSUE TITLE
Small improvement for "Dockerizing a Node.js web app" guide

### ISSUE THREAD

**Issue by fabriziocucci** [nodejs/nodejs.org#1186]
* **https://nodejs.org/en/docs/guides/nodejs-docker-webapp/**:
* **Browser version not relevant**:
* **Operating system not relevant**:

It would be nice to extend the guide by including the motivation for splitting the copy step in two:

```docker
# Install app dependencies
COPY package.json /usr/src/app/
RUN npm install
```

```docker
# Bundle app source
COPY . /usr/src/app
```

I guess a quick explanation that mentions cached layers (similar to [this](http://bitjudo.com/blog/2014/03/13/building-efficient-dockerfiles-node-dot-js/)) would be much appreciated, especially by Docker beginners.

**Comment by sotayamashita** [comment]
@fabriziocucci I agree with you. we should create a section such as "conclusion" and it show the whole code which we need to build node app on docker. what do you think?

**Comment by fabriziocucci** [comment]
> and it show the whole code which we need to build node app on docker

@sotayamashita are you referring to what's already at the end of the section [Creating a Dockerfile](https://nodejs.org/en/docs/guides/nodejs-docker-webapp/#creating-a-dockerfile) or something else?

What (I think) I was referring to when I opened this issue is that it would be nice to mention **why** we write in the Dockerfile:

```docker
# Install app dependencies
COPY package.json /usr/src/app/
RUN npm install

# Bundle app source
COPY . /usr/src/app
```

instead of:
```docker
# Bundle app source
COPY . /usr/src/app

# Install app dependencies
RUN npm install
```

Unless the reader is already aware of the Docker cached layers (as explained [here](http://bitjudo.com/blog/2014/03/13/building-efficient-dockerfiles-node-dot-js/)), copying only the `package.json`, calling `npm install` and then copying the whole content of the current directory may seem like a twisted alternative to the second option.

Does that make sense?

**Comment by fhemberger** [comment]
@fabriziocucci Great idea, would you mind giving it a try and submit a PR for this?

**Comment by marxenegls** [comment]
praise Krishna & fabriziocucci ..
i have wrecked my small busy head for days .... 
WHY HASN'T the tutorial been updated to reflect the changes on Node.js main site no less? 
'https://nodejs.org/en/docs/guides/nodejs-docker-webapp/'
It was severely discouraging to spend days failing to simply get node to say hello , only through my sheer perseverance have i stuck it out till i found this post . I've written my own 'issue' but couldn't bring myself to post it . i ought to mention it if only so that no one else gets stuck on the same issue, otherwise there'll be little point for such a site.

Step 3/7 : COPY package.json package-lock.json .
COPY failed: stat /var/lib/docker/tmp/docker-builder682146220/package.json: no such file or directory

**Comment by maximlomans** [comment]
please consider editing the tutorial / Doc on the Node.js site 'Dockerizing Node.js ...' (https://nodejs.org/en/docs/guides/nodejs-docker-webapp/)
what fabriziocucci  suggested hasn't yet happened , i just spent two days beating myself up as to why following a tutorial led to so many errors.
 Install app dependencies
COPY package.json /usr/src/app/
RUN npm install

 Bundle app source
COPY . /usr/src/app
thank you

**Comment by fhemberger** [comment]
@marxenegls Yes, there is a bug in the tutorial (I just fixed it in #1316), but it's completely unrelated to this issue.
